Add support for littlefs

pabigot opened the pull request:
[littlefs](https://github.com/ARMmbed/littlefs) is a wear-leveling file system with power-loss resilience and bounded memory use, suitable for use on Zephyr flash partitions including SPI NOR devices and SOC flash devices.

This commit, based on an initial implementation provided by @jimparis, adds support for the file system, with an example and shell support.

pabigot committed:
subsys/fs: add support for littlefs

littlefs is a fail-safe filesystem from ARM Mbed that has wear-leveling
capabilities.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>
Signed-off-by: Jim Paris <jim@bolt.io>

pabigot committed:
samples/subsys/fs/littlefs: add a basic sample

Uses a littlefs file system to maintain a boot counter.  Also tests some
other functions.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
subsys/fs/shell: add littlefs support

Add support for the littlefs file system in the fs shell.  Update
the sample to use the same partition configuration as the littlefs
example for the SPI NOR test platform.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

galak commented on west.yml:
```diff
@@ -25,6 +25,8 @@ manifest:
     - name: upstream
       url-base: https://github.com/zephyrproject-rtos
+    - name: pabigot
+      url-base: https://github.com/pabigot
```
I assume this is until we host littlefs module in Zephyr project?

pabigot replied:
Yes, @carles is creating that today.

galak reviewed (commented):
did you mean to have a merge commit in the PR?

pabigot commented:
> did you mean to have a merge commit in the PR?

Yes, see the description at top.

pabigot committed:
subsys/fs/shell: avoid warning about output format truncation

The previous way of creating an absolute path relied on snprintf(), and
when used with newlib gcc warned that the output could be truncated
before the last format character.  Rework to use code that doesn't rely
on snprintf.

See discussion at https://bugzilla.redhat.com/show_bug.cgi?id=1431678

Also ensure that cwd is always NUL-terminated, and use the utility
function to create the absolute path in cmd_trunc.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented on samples/subsys/fs/littlefs/README.rst:
```diff
@@ -0,0 +1,69 @@
+Building and Running
+********************
+
+This example should work on any board that defines a ``storage``
```
Yes.  https://docs.zephyrproject.org/latest/guides/bluetooth/bluetooth-arch.html#persistent-storage has it, so it should be "storage" rather than `storage`.  OK?

dbkinder replied:
I found that in my grepping, (and found a bunch of .overlay files defining partitions with the "storage" label, but I didn't find anything specific of how to define a "storage" partition.  Oh well, maybe it's obvious to those that know.

pabigot replied:
Please see whether the rework reads better.  The use in Requirements is more complete; the use here is a basic reference.

pabigot commented on samples/subsys/fs/littlefs/README.rst:
```diff
@@ -0,0 +1,69 @@
+.. _sample_subsys_fs_littlefs:
+
+littlefs Filesystem Sample Application
```
Will do, though littlefs spells it without the space, and 30% of the Zephyr uses do too, and back in the day that was the common usage, so I'll probably recidivate.

pabigot replied:
so does the fat_fs example I copied from.

pabigot replied:
done

dbkinder replied:
BTW, I was just looking at consistency in this document.  There were multiple uses of "file system" (with a space) and only this one use of filesystem in the title (without a space).  And yes, sigh, being consistent across all the documentation would be nice. :)

pabigot commented on samples/subsys/fs/littlefs/README.rst:
```diff
@@ -0,0 +1,69 @@
+********************
+
+This example should work on any board that defines a ``storage``
+partition.  Two specific targets are tested.
```
done

pabigot commented:
Addressed public reviews to this point , and non-public related to handling of `fs_data`.  Also added a change to `fs/shell` required to avoid build warnings using newlib.

dbkinder reviewed (approved):
doc changes LGTM, thanks!

pabigot commented:
Close and reopen to restart shippable now that zephyr's littlefs repo has content.

pabigot commented:
I'd like to point out that, for historical reasons, the default flash page layout block size is 64 KiBy, which is huge.  For littlefs, and probably most other file systems, a more sane setting is 4 KiBy.  You get this by setting `CONFIG_SPI_NOR_FLASH_LAYOUT_PAGE_SIZE=4096`.

tyrantlizard77 reviewed (commented):
I'm not seeing any tests which prove that this works, only samples.

Can you either convert some of these samples into actual tests, or amend current fs subsystem tests to run with littlefs as a backend?

pabigot commented:
I looked into cloning the NFFS tests which are comprehensive but also appear to be whitebox so non-trivial to port to a different file system.  I'll revisit this next week.

vanwinkeljan commented on samples/subsys/fs/littlefs/sample.yaml:
```diff
@@ -0,0 +1,7 @@
+tests:
+  sample.subsys.fs.littlefs:
+    build_only: true
+    platform_whitelist: particle_xenon
```
Readme mentions 2 tested platforms `nrf52840_pca10056 ` and `particle_xenon` and here only one of them is whitelisted, any reason for that?

pabigot replied:
I think PCA10056 was added later just because it's more common.  I can put both there.

vanwinkeljan commented on subsys/fs/CMakeLists.txt:
```diff
@@ -4,18 +4,21 @@ if(CONFIG_FILE_SYSTEM)
   zephyr_interface_library_named(FS)
   #zephyr_link_interface_ifdef(CONFIG_FAT_FILESYSTEM_ELM ELMFAT)
   #zephyr_link_interface_ifdef(CONFIG_FILE_SYSTEM_NFFS   NFFS)
+  #zephyr_link_interface_ifdef(CONFIG_FILE_SYSTEM_LITTLEFS LITTLEFS)
```
Any idea why all this `zephyr_link_interface_ifdef ` are commented out?

pabigot replied:
Looks like @nashif disabled them when moving the previous file systems to a module.  I can remove them in a second commit.

vanwinkeljan commented on subsys/fs/Kconfig:
```diff
@@ -39,6 +39,13 @@ config FILE_SYSTEM_NFFS
+config FILE_SYSTEM_LITTLEFS
+	bool "LittleFS file system support"
+	depends on FLASH_MAP
+	depends on FLASH_PAGE_LAYOUT
```
According to the samples littlefs depends on newlib, so I would expect a `depends on NEWLIB_LIBC` here

pabigot replied:
littlefs depends on an implementation of `strspn` and `strcspn`.  newlib is not the only provider of that capability, it's just that minimal libc doesn't have it.

Probably the right answer is to implement those two functions for minimal libc, but that's not in scope right now.

vanwinkeljan commented on subsys/fs/shell.c:
```diff
@@ -481,14 +507,52 @@ static int cmd_mount_nffs(const struct shell *shell, size_t argc, char **argv)
+
+static int cmd_mount_littlefs(const struct shell *shell, size_t argc, char **argv)
+{
+	printk("mount littlefs at %s\n", argv[1]);
```
`shell_print` should be used instead of `printk`

vanwinkeljan commented on samples/subsys/fs/littlefs/src/main.c:
```diff
@@ -0,0 +1,140 @@
+void main(void)
+{
+	struct fs_mount_t *mp = &lfs_storage_mnt;
+	unsigned int id = (uintptr_t)mp->storage_dev;
```
Should this not be a cast `unsigned int`?

pabigot replied:
No.  It's only valid to cast `void *` to or from `[u]intptr_t`: direct cast to `unsigned int` would be incorrect if pointers require a larger underlying type.

There could be an explicit cast to truncate to the destination type, though it's not required by C for unsigned integral types.

vanwinkeljan commented on subsys/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,666 @@
+		 *
+		 * In v2.1 zero is expected to be disallowed and
+		 * leveling will be disabled by using -1 (see issue
+		 * #233).  Wait until that happens before implementing
```
Maybe mention that the issue lives in ARMmbed/littlefs

pabigot committed:
subsys/fs: clean up CMakeLists

Remove the disabled link interface commands.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
@vanwinkeljan Please revisit your review.  I've marked resolved the issues that I've addressed explicitly.  The one on the pointer-to-int cast is left open for any further discussion.

For newlib I'll create a separate commit that extends minimal libc.  If it gets approved and merged before this I'll remove the dependency in the samples.

vanwinkeljan commented:
Reference to minimal libc strspn/strcspn PR #17817

pabigot commented:
Updated upstream version to v2.1.0 and added myself as a codeowner of the main module.

nvlsianpu commented on samples/subsys/fs/littlefs/prj.conf:
```diff
@@ -0,0 +1,24 @@
+CONFIG_MAIN_STACK_SIZE=2048
+
+CONFIG_LOG=y
+#CONFIG_FS_LOG_LEVEL_DBG=y
```
remove commented line

pabigot replied:
It's there as a hint so people don't waste time trying to set `CONFIG_FILE_SYSTEM_LOG_LEVEL_DBG=y` to figure out what's going wrong.

nvlsianpu replied:
ok, maybe worth to add appropriate comment (up to you)

nvlsianpu reviewed (commented):
great contribution - this is in line with what we were considered to port for zephyr some time ago. I will review this more deeply soon

carlescufi commented:
> I'm not seeing any tests which prove that this works, only samples.
> 
> Can you either convert some of these samples into actual tests, or amend current fs subsystem tests to run with littlefs as a backend?

I think the latter would be a good starting point. The [upstream tests](https://github.com/ARMmbed/littlefs/tree/master/tests) all seem to run on desktop only and we make no real changes to the upstream code, so reusing existing FS tests running on this particular filesystem is the easier and safer option for us in my opinion.

pabigot commented:
OK, so I agree this needs tests.  I would never have thought that a `fs_rename(from, to)` function would require recursively removing `to` if it already exists.  Rather than returning an error and keeping your data safe from your mistake.

This'll take a while: none of the existing filesystem tests are complete (fatfs) or generic (nffs) enough to be reused without significant rework.

pabigot committed:
tests/subsys/fs/littlefs: add tests

Verify all FS API calls using the nRF52840 development kit.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
Latest push adds tests, fixes a few bugs including an upstream patch, and removes the newlib dependency.  IMO this is ready for final reviews.

pkral78 commented on subsys/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,663 @@
+	fs_lock(fs);
+
+	struct lfs_info info;
+	int ret = lfs_dir_read(&fs->lfs, dp->dirp, &info);
```
I've just noticed that littlefs add by default '.' and '..' directory entries (and it's not configurable in LTS module). This is opposite to FatFS and NFFS behaviour. We can remove them here or leave decisions on higher layers.

pabigot replied:
Yeah.  I'm pretty ambivalent about it.  I don't know about NFFS, but I don't think FatFS generates those entries (later: apparently it can); if neither of their wrappers are filtering the results, there's an argument that littlefs shouldn't either.

Here's the upstream rationale for their presence: https://github.com/ARMmbed/littlefs/issues/175#issuecomment-493103440

On the other hand the entries really aren't very useful (they're always the first two returned from readdir).  If existing users of the Zephyr FS API are not prepared to handle them, it'd be better to remove them at the wrapper layer (and document in the API that they are not to be provided by implementations).

Absent other comment I'll go that route.  Thanks for bringing this up.

pabigot replied:
>  better to remove them at the wrapper layer (and document in the API that they are not to be provided by implementations).

Except I just realized that this file system may be used under the POSIX wrappers, where they would be expected and handled properly.   @jimparis @pfalcon any comment?

Whichever way this goes the rationale needs to be documented:
* `fs_readdir` returns all entries supplied by the underlying file system implementation; or
* `fs_readdir` strips `"."` and `".."` for consistency with other file system implementations.

pabigot replied:
After looking at it I think the preferred solution is outside the scope of littlefs, so I'm punting this to #17951, and will keep the current approach of not filtering littlefs results.

vanwinkeljan replied:
@pabigot keep in mind that the FUSE fs access layer is adding them to be posix compliant.

https://github.com/zephyrproject-rtos/zephyr/blob/af9a2670d254a865faa6bff4261995616a23f48a/subsys/fs/fuse_fs_access.c#L203-L204

pabigot replied:
Sounds like FUSE is doing that because the FS abstraction isn't well defined, and option 2 of #17951 is the right way to go.

pkral78 commented on subsys/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,663 @@
+	BUILD_ASSERT((CONFIG_FS_LITTLEFS_CACHE_SIZE
+		      % CONFIG_FS_LITTLEFS_PROG_SIZE) == 0);
+
+	lfs_size_t read_size = CONFIG_FS_LITTLEFS_READ_SIZE;
```
Another thing to consider. This Kconfig static configuration disqualifies use cases using littlefs for multiple flash devices with different read/prog properties (i.e. internal MCU flash, and an external one). 

Could be these properties defined somewhere in DTS?

pabigot replied:
That's not really true: these settings affect only in-memory caches and alignment of I/O transactions.  As long as the read and prog sizes are a multiple of the flash read/write alignment requirements (1 for SPI-NOR, 4 generally for internal flash) there's no conflict.  The defaults are selected to reduce RAM requirements.

I did look into providing these through a config structure associated with the storage device, with an eye to supporting bigger sizes for partitions with larger data files.  The complexity of managing/validating/using the configuration was high, and limited experimentation didn't really show significant performance advantages to increasing them beyond the default.  That can be revisted in the future, keeping the Kconfig values as defaults when there is no mount-point--specific override.

These properties really are application relevant and belong in Kconfig.  They can be changed without affecting the ability to use an existing file system, unlike the size of the erase block (if that changes, the filesystem will be reformatted).

The erase block comes from devicetree via the flash layout API and can vary between devices.  I have tested littlefs on both SOC flash and SPI-NOR--based file systems using these values.  You can also have partitions on both flash technologies in the same application.

I don't see a great value in changing the approach at this time.

pkral78 replied:
> That's not really true: these settings affect only in-memory caches and alignment of I/O transactions. As long as the read and prog sizes are a multiple of the flash read/write alignment requirements (1 for SPI-NOR, 4 generally for internal flash) there's no conflict. The defaults are selected to reduce RAM requirements.
> 
> I did look into providing these through a config structure associated with the storage device, with an eye to supporting bigger sizes for partitions with larger data files. The complexity of managing/validating/using the configuration was high, and limited experimentation didn't really show significant performance advantages to increasing them beyond the default. That can be revisted in the future, keeping the Kconfig values as defaults when there is no mount-point--specific override.

Even if you were right in most cases, I don't see this kind of inflexibility in configuration API as right way to go.  Moreover, what's the benefit of storing 

```c
struct lfs_config cfg;
```
in RAM ? It has about 72 bytes. For most configurations it should be known at compile time and could be placed in flash. However, there should also be way to create it in runtime of course. The most simple solution would be to introduce config macros and allows to predefine it in Kconfig (for static configurations).

pabigot replied:
So, as I said, I *tried* this and it isn't as easy as you think once you get into the details.

However, I share your concerns that at some point in the future per-filesystem configurability might be worthwhile, and the way it will have to be done would require a change to the way applications define littlefs partitions.

OK.  I'll take another stab at it.

pabigot replied:
This is done.  It's not too painful to use the defaults, though it's necessarily different from all the other file systems.

There seems to be a bug in littlefs, related to ARMmbed/littlefs#224, where using a `cache_size` of 512 or larger causes an infinite loop when trying to write a block of data larger than the cache size to an empty file: it writes one cache block of data to each filesystem block until it runs out of disk space.  Apparently the code that transfers from inline files (ones with less than 512 bytes of content) to outline files mishandles this.

As such, the maximum I/O transaction size that works is 256, and the performance improvement from non-default configuration is not great, at least not with a SPI NOR back-end.

pabigot committed:
subsys/fs: remove ambiguity in readdir results

Existing file system implementations do not provide the special "."
(current) and ".." (parent) directory entries in the readdir results.
littlefs does.

Remove these entries in the abstraction layer.  This simplifies code in
higher level consumers that aren't prepared to see them.  Consumers like
FUSE that need them can put them back without having to worry about
conflicts.

Closes issue #17951

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
subsys/fs/littlefs: allow customization of file system configuration

There's desire to be able to customize parameters on a per-filesystem
basis, which means we need a way to override the Kconfig defaults which
are global.  This also means the littlefs data structure cannot own the
cache and lookahead buffers.

Switch to using a macro to define the littlefs data structure.  The
default version uses the Kconfig constants.  A custom one takes
arguments providing the most likely partition-specific parameters.
Finally the user is free to bypass the helper macros and set any
parameters desired, though validation is limited and only present when
CONFIG_DEBUG is enabled.

Extend the test suite with a performance module, which confirms that
these settings have an impact proportional to the log of changes to the
cache or IO sizes.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
This is getting bigger and bigger.

Rebased on top of #17994 which fixes a serious correctness bug in writes that should cross a page boundary, and a serious performance problem with reads.  That PR needs reviewers.

This now also includes a fix for #17749.  @vanwinkeljan note for FUSE/POSIX (you shouldn't have to do anything).

zephyrproject-rtos/littlefs has been updated with a couple bug fixes; the ones present in upstream have been submitted (ARMmbed/littlefs#253, ARMmbed/littlefs#254).

The initial support for per-filesystem customization missed the need for a per-file data cache that matches the filesystem cache size.  Since the original implementation hard-coded the default size, this didn't work.  There's now a slab when using default-only partitions, and memory pool for full customization.  Examples of the different ways to configure your file system are in the test area.

Here's the performance test results for the Particle Xenon's 4 MiBy flash.  After all the bug fixes there is some benefit to using larger cache sizes.  Dunno how it compares to other solutions, but it'll meet my needs.

```
starting test - test_lfs_perf
clearing /sml for small 8x1K dflt write/read test
Erasing 65536 (0x10000) bytes
Wiped flash area 4 for /sml
/sml: bsize 16 ; frsize 4096 ; blocks 16 ; bfree 14
read_size 16 ; prog_size 16 ; cache_size 64 ; lookahead_size 32
creating and writing 8 1024-byte blocks
small 8x1K dflt write 8 * 1024 = 8192 bytes in 236 ms: 34711 By/s, 33 KiBy/s
small 8x1K dflt read 8 * 1024 = 8192 bytes in 49 ms: 167183 By/s, 163 KiBy/s
bufs 0x200043dc 0x20004bdc 0x200053dc
clearing /sml for small 8x1K bigfile write/read test
Erasing 65536 (0x10000) bytes
Wiped flash area 4 for /sml
/sml: bsize 256 ; frsize 4096 ; blocks 16 ; bfree 14
read_size 256 ; prog_size 256 ; cache_size 1024 ; lookahead_size 128
creating and writing 8 1024-byte blocks
small 8x1K bigfile write 8 * 1024 = 8192 bytes in 176 ms: 46545 By/s, 45 KiBy/s
small 8x1K bigfile read 8 * 1024 = 8192 bytes in 36 ms: 227555 By/s, 222 KiBy/s
clearing /med for medium 32x2K dflt write/read test
Erasing 983040 (0xf0000) bytes
Wiped flash area 5 for /med
/med: bsize 64 ; frsize 4096 ; blocks 240 ; bfree 238
read_size 64 ; prog_size 64 ; cache_size 256 ; lookahead_size 64
creating and writing 32 2048-byte blocks
medium 32x2K dflt write 32 * 2048 = 65536 bytes in 1283 ms: 51080 By/s, 49 KiBy/s
medium 32x2K dflt read 32 * 2048 = 65536 bytes in 285 ms: 229950 By/s, 224 KiBy/s
clearing /lrg for large 64x4K dflt write/read test
Erasing 3145728 (0x300000) bytes
Wiped flash area 6 for /lrg
/lrg: bsize 256 ; frsize 32768 ; blocks 96 ; bfree 94
read_size 256 ; prog_size 256 ; cache_size 1024 ; lookahead_size 128
creating and writing 64 4096-byte blocks
large 64x4K dflt write 64 * 4096 = 262144 bytes in 4517 ms: 58034 By/s, 56 KiBy/s
large 64x4K dflt read 64 * 4096 = 262144 bytes in 981 ms: 267221 By/s, 260 KiBy/s
PASS - test_lfs_perf
```

The nrf52840_pca10056 is slower in writes, and faster in reads.

pabigot commented:
recheck build after updating littlefs fork

vanwinkeljan commented:
> This now also includes a fix for #17749. @vanwinkeljan note for FUSE/POSIX (you shouldn't have to do anything).

Noticed the change and it looks good to, only one remark the issue number should read #17951

pabigot commented:
Latest push corrects the issue number in a comment and switches to consistently using memory pool for file cache allocations, which simplifies the code and increases flexibility for minor cache size customizations.

pkral78 commented on include/fs/littlefs.h:
```diff
@@ -0,0 +1,106 @@
+/** @brief Filesystem info structure for LittleFS mount */
+struct fs_littlefs {
+	/* Defaulted in driver, customizable before mount. */
+	struct lfs_config cfg;
```
Would it be possible to change this to:
```c
const struct lfs_config *cfg;
```

and

1) extend configuration macros so it will be possible to setup `lfs_config` as `static const` structure at compile time and has it whole at a flash
2) add function

```c
int littlefs_config(struct fs_mount_t *mountp, struct lfs_config *cfg);
```
which will be runtime version (i.e. need to be called before `littlefs_mount`) ?

pabigot replied:
For (1), no.  `lfs_config.{read,prog,sync}` are assigned within the driver to the functions the driver implements for a specific back-end.  The structure must be in RAM.

For (2), I don't understand who would be calling that function.  It's not in the FS API.

pkral78 replied:
> For (1), no. `lfs_config.{read,prog,sync}` are assigned within the driver to the functions the driver implements for a specific back-end. The structure must be in RAM.
> 
I don't see any specific back-end other than defined in lfs_api_xxx functions (i.e. functions that wraps flash_area_xxx functions) and even if there would be so, it certainly does not need to be in RAM.

> For (2), I don't understand who would be calling that function. It's not in the FS API.

Every concrete filesystem in Zephyr need specific setup and initialization before mount (for example in fs shell, there are mount functions for fat, nffs and littlefs. Those initialization API does not need to be FS API, it is specific to FS (so better than API is to use term SPI in meaning of 'service provider interface').

From use cases point of view there are two:
1.) static FS configuration - everything is known before build, i.e. flash parameters, sizes, offsets, buffers and other littlefs settings - this is common for most cases and target devices.
2.) runtime FS configuration that otherwise need to call second function - for example posix host utility configured by arguments from command line, or dynamic FS setup for micropython etc.

I just don't like to see wasting RAM for constant structures.

pabigot replied:
It is not a constant structure given the way the FS API works.  See https://github.com/zephyrproject-rtos/zephyr/pull/17749/files#r310552307.

The shell really does nothing more than put some generic glue code that's specific to the shell around a call to `fs_mount()` given a pointer to a statically-defined mount point.  NFFS has to get the device name at runtime; fatfs and littlefs do not.  Generally there is no FS-specific runtime configuration.

pkral78 replied:
> It is not a constant structure given the way the FS API works. See https://github.com/zephyrproject-rtos/zephyr/pull/17749/files#r310552307.
> 
> The shell really does nothing more than put some generic glue code that's specific to the shell around a call to `fs_mount()` given a pointer to a statically-defined mount point. NFFS has to get the device name at runtime; fatfs and littlefs do not. Generally there is no FS-specific runtime configuration.

For static configurations you wouldn't be forced to call littlefs specific config function. However there are other cases where use of such interface is legitimate: format, chkdsk, etc.) without it, it would not even be possible to create utility like described in #17719 (without recompilation everytime FS parameters is changed).

pabigot replied:
So in those special cases you need runtime configuration, which you can do with this as it stands.

It seems that all this is about trying to move 72 bytes of required configuration data from RAM into flash.  As the submitter and maintainer of this module my answer is: no.  It doesn't increase functionality, and it dramatically increases complexity because of the visibility and accessibility of information that would have to be collected at compile time.

It's up to you whether you want to approve, comment, or reject this PR.  If you want to figure out how to make this work later, that's on you.

pfalcon replied:
> figure out how to make this work later

I'd definitely recommend any reviewer to keep this option in mind. Suggesting an improvement may be good, but an answer "it can't be all done at once" should be a well expected response.

pkral78 replied:
> So in those special cases you need runtime configuration, which you can do with this as it stands.
> 
> It seems that all this is about trying to move 72 bytes of required configuration data from RAM into flash. As the submitter and maintainer of this module my answer is: no. It doesn't increase functionality, and it dramatically increases complexity because of the visibility and accessibility of information that would have to be collected at compile time.
> 
> It's up to you whether you want to approve, comment, or reject this PR. If you want to figure out how to make this work later, that's on you.

Yeah, its not my intention to stall this PR. As I said I agree with it as is.

pabigot replied:
OK.  When the review is comment-only it's hard to tell where change requests fall on the spectrum from "it would be nice if..." to "if you don't change this I'll block it".

I usually try pretty hard to accommodate requests (I spent an hour trying to write up why I couldn't make partitions independently configurable before I figured out how to make it work), but this one just isn't going to work at this time.

@pkral78 if github lets you select "Approve" in the review box, or change the review type from "Comment" to "Approve", please do.  Comment-only approvals are nice, but the folks who merge won't see them.  If it won't let you do that, ok.  Thanks for your input; it's improved the solution significantly.

pabigot commented on subsys/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,730 @@
+	lcp->prog = lfs_api_prog;
+	lcp->erase = lfs_api_erase;
+	lcp->sync = lfs_api_sync;
+	lcp->read_size = read_size;
```
@pkral78 Here is where `lfs_config` must be mutable.  The pointers to the device-private functions get assigned, as well as information about where in the flash device the specific partition is located.

pkral78 replied:
No it does not need to, those are just immutable function pointers or constants known at compile time. I see block size is get from `get_block_size` call, however I don't understand why it need to be calculated at runtime, when it will be always same for same target configuration.

pabigot replied:
`prog`, `erase`, and `sync` are known at compile time, but they're private to this module.  How do you expect to be able to set them at compile time in const struct stored in flash and declared in a different module?

The flash characteristics are determined at runtime because Zephyr's devicetree implementation only gives you the flash area identifier.  The remaining information simply isn't available readily at compile time.

pkral78 replied:
> `prog`, `erase`, and `sync` are known at compile time, but they're private to this module. How do you expect to be able to set them at compile time in const struct stored in flash and declared in a different module?
> 

And does need to be private ? 

> The flash characteristics are determined at runtime because Zephyr's devicetree implementation only gives you the flash area identifier. The remaining information simply isn't available readily at compile time.

Yes you're right, but it's rather DT limitation (an area for improvement?).

I think that these two questions lead to decision if we could accept this PR as is and left my comments for future improvements. I am for accepting this PR as it is super useful and well written.

pabigot commented:
Rebased on merged spi nor fixes, removed redundant target support in test.

jimparis commented:
Looking good, thanks for all your work.  I haven't had time to do a deep review, but I've started upgrading my local tree to a newer zephyr + your branch.

One build issue: undefined `__ASSERT_NO_MSG` in `lfs.c` from `lfs_util.h`.  I'm using `CONFIG_NEWLIB_LIBC` and newlib's `assert.h` doesn't have that zephyr-specific function.

nvlsianpu commented on subsys/fs/fs.c:
```diff
@@ -255,11 +255,25 @@ int fs_readdir(struct fs_dir_t *zdp, struct fs_dirent *entry)
+					break;
+				}
+				if (entry->type != FS_DIR_ENTRY_DIR) {
+					break;
```
Shoudn't this case returns an errcode?

pabigot replied:
No.  If it's not a directory entry, it's a candidate for return (probably a file, maybe something else).  This just allows fall through to the subsequent check for names of special directories, the last condition for whether something gets filtered out.

nvlsianpu commented on subsys/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,730 @@
+	case LFS_ERR_IO:        /* Error during device operation */
+		return -EIO;
+	case LFS_ERR_CORRUPT:   /* Corrupted */
+		return -EIO;
```
Can this be other errorcode than -EIO? -EIO should be reserved for low level raw data access errors (like flash driver/map can't access a data block etc.).

nvlsianpu replied:
maybe -EFAULT?

pabigot replied:
Switched to `-EFAULT`.  It's no more right than anything else (`LFS_ERR_CORRUPT` is used in a variety of places in littlefs, some of which map to `-EINVAL` or `-EIO`), but it does make it possible to uniquely identify the LFS-layer error which is important.

nvlsianpu commented:
@pabigot PR looks fine to me, it is tremendous job. Please answer to my comments, and I will accept it.

carlescufi commented:
CC @geky, in case you are interested in this integration

pabigot commented:
> One build issue: undefined `__ASSERT_NO_MSG` in `lfs.c` from `lfs_util.h`. I'm using `CONFIG_NEWLIB_LIBC` and newlib's `assert.h` doesn't have that zephyr-specific function.

Updated in the Zephyr fork and this PR.  (Weird; something must have changed in zephyr, since originally we had to use newlib for this.)

geky commented:
Oh! This is very cool. Thanks for the putting this together @pabigot.

I'm going to have to add reference to this in littlefs if that's cool with you guys.

pabigot commented:
> Oh! This is very cool. Thanks for the putting this together @pabigot.

You're welcome; and again thanks to @jimparis for the starting point.
>
> I'm going to have to add reference to this in littlefs if that's cool with you guys.

I think that should be fine; @carlescufi  is there any specific phrasing we would like from a project perspective?

carlescufi commented:
> > Oh! This is very cool. Thanks for the putting this together @pabigot.
> 
> You're welcome; and again thanks to @jimparis for the starting point.
> 
> > I'm going to have to add reference to this in littlefs if that's cool with you guys.
> 
> I think that should be fine; @carlescufi is there any specific phrasing we would like from a project perspective?

@geky @pabigot, no, since this is text to be added to an external project all that I could ask for is that Zephyr to be spelled correctly :)
Thanks!
